docs(serverless): secure your function/container (#1258)
scaleway/docs-content: serverless/functions/how-to/secure-a-function.mdx

---
meta:
  title: How to secure a function
  description: This page explains how to secure a Serverless Function
content:
  h1: How to secure a function

>   paragraph: This page explains how to  how to secure a Serverless Function

Reviewer: ```suggestion
  paragraph: This page explains how to secure a Serverless Function
```

tags: functions secure private
dates:
  validation: 2023-02-01
  posted: 2023-02-01
categories:
  - serverless
---

This page explains how to secure your function.

<Macro id="iam-requirements" />

<Message type="requirement">
  - You have an account and are logged into the [Scaleway console](https://console.scaleway.com)
  - You have [created a functions namespace](/serverless/functions/how-to/create-a-functions-namespace/)
  - You have [created a Function](/serverless/functions/how-to/create-a-function/)
</Message>

## Use Secrets to store access keys and sensible information

Instead of using environment variables (which are stored in clear) for your functions, use Secrets. These are information usable through environment variables but are stored encrypted. 

### Configure Secrets from the Scaleway Console

1. Click **Functions** in the **Serverless** section of the side menu. The functions page displays.

> 2. Click the relevant functions namespace. 

Reviewer: ```suggestion
2. Click the relevant function namespace. 
```

3. Click the name of the function for which you want to define Secrets.
4. Click the **Deployment** tab.
5. Scroll to the **Secrets** section of the page and click **Add a new secret**. Enter the key and value for your Secret. Repeat for additonal secrets.
    <Message type="important">
        Be careful when you type your Secrets. You will not be able to read the value in the console once submitted.
    </Message>
6. Click **Deploy a Function** to submit your Secrets and redeploy your function. 

### Configure Secrets using the Serverless framework

> Add `secret` to your functions description (more information in the [plugin documentation](https://github.com/scaleway/serverless-scaleway-functions#security-and-secret-management))

Reviewer: ```suggestion
Add `secret` to your function's description (more information in the [plugin documentation](https://github.com/scaleway/serverless-scaleway-functions#security-and-secret-management))
```

> We recommend to use them with global environment variables or a `.env` file stored independently (and kept secret).

Reviewer: ```suggestion
We recommend using them with global environment variables or a `.env` file stored independently (and kept secret).
```

``` 
secret:
      secret_1: ${env:SCW_SECRET_KEY} 
```
<Message type="important">

>     We recommend you to not commit in a Version Control System (VCS), and to not share your Project ID or access key to ensure the security of your configuration file, which may contain sensitive data.

Reviewer: as above

>     We recommend you to not commit in a Version Control System (VCS), and to not share your Project ID or access key to ensure the security of your configuration file, which may contain sensitive data.

Author: ```suggestion
    We recommend you to not commit in a Version Control System (VCS), and to not share your Project ID or access key to ensure the security of your configuration file, which may contain sensible data.
```

>     We recommend you to not commit in a Version Control System (VCS), and to not share your Project ID or access key to ensure the security of your configuration file, which may contain sensitive data.

Reviewer: ```suggestion
    We strongly suggest that you do not commit this in a Version Control System (VCS), and do not share your Project ID or access key. This helps to ensure the security of your configuration file, which may contain sensitive data.
```

</Message>

### Configure Secrets using Terraform

Add the following [resource description](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs/resources/function) in Terraform:

```
secret_environment_variables = { "key" = "secret" }
```

## Restrict access to your functions

You can set Serverless Functions as private if you want to protect your functions from unwanted or unauthorized calls. 

> Unauthenticated calls will be rejected, and your function will not be triggered. This feature is handy if an event triggers your function (CRON, and soon SQS or NATS trigger) or if you put them behind an API gateway or a proxy server (see examples in serverless-examples).

Reviewer: should there be a link to serverless-examples?

> Unauthenticated calls will be rejected, and your function will not be triggered. This feature is handy if an event triggers your function (CRON, and soon SQS or NATS trigger) or if you put them behind an API gateway or a proxy server (see examples in serverless-examples).

Author: ```suggestion
Unauthenticated calls will be rejected, and your function will not be triggered. This feature is handy if an event triggers your function (CRON, and soon SQS or NATS trigger) or if you put them behind an API gateway or a proxy server (see examples in [serverless-examples](https://github.com/scaleway/serverless-examples)).
```

### Restrict access from the Scaleway Console

1. Click **Functions** in the **Serverless** section of the side menu. The functions page displays.

> 2. Click the relevant functions namespace. 

Reviewer: ```suggestion
2. Click the relevant function namespace. 
```

3. Click the name of the function for which you want to define Secrets.
4. Click the **Security** tab.
5. Set the **Privacy Policy** of the function to **Private**. 
6. If required create an access token for your function:
    * Click **Generate token** in the **Tokens** section of the page. A pop-up displays.

>     * Enter a description and set an optional expiration date. To have a permanent valid token tick **No expiration date**. Then click **Generate token**.

Reviewer: ```suggestion
    * Enter a description and set an optional expiration date. To generate a permanently-valid valid token tick **No expiration date**. Then click **Generate token**.
```
